### DISCUSSION TITLE
Reduced Eye Model

### DISCUSSION THREAD

**Discussion by FERpo-23** [mjhoptics/ray-optics#124] [General]
Hi, I really appreciate your effort to develop the package.
One of my colleagues told me about it and I downloaded to take a look.
The main goal for me was to use it as an educational resource in my teaching activities, instead of using other commercial computer resources.
To this end I took a look to the python notebook related to Issue 92.

I made a very simple model of an eye (the so called reduced eye) with a single spherical surface, with air at the left hand side and water (refractive index of 'n') to the right hand side.
The aperture stop in this very simple model is located at the vertex of the refracting surface, thus the entrance and exit pupil are located at that place and share its size.

1. In working with this simple model I realized that the paraxial calculation provides values for the image focal length that had to be multiplied by 'n'. The same happens when computing the location of the exit pupil0.
2.  I also obtained a Ray Spot Diagram, and I do not understand the reason for the large values of transverse aberrations at 
+- 45 degrees.

Any indication will be acknowledged.
Regards.
Fernando
[Reduced_eye_1S.tar.gz](https://github.com/mjhoptics/ray-optics/files/12302212/Reduced_eye_1S.tar.gz)

PS: I enclose a tar-gz file with the Ipython notebook for it could help to clarify the questions

**Comment by mjhoptics** [comment]
Hello @FERpo-23,
Thank you for your report. There was a missing factor of the image space refractive index in the efl calculation. I have updated a fix to the master branch with this correction.
I also see a potential source of confusion about what ray-optics considers image space. The image space gap and it's medium is where the first order calculations look for this index value. When you added the second planar surface, that created a zero thickness air gap before the image plane. The model can be corrected in one of several ways:

1. change the medium of the final gap to be the Medium Eye.
```
#units in 'mm'
sm.gaps[0].thi=1e12#distant object
sm.add_surface([Rco, rp, n, 56., 1.5]) #This provides the surface representing the eye with a fake V-number
sm.gaps[sm.cur_surface].medium = med2 = ConstantIndex (n, 'Medium Eye') 
#---
sm.add_surface([0., 0.])                #dummy surface-->0.0--->Simulate a plane retina
sm.ifcs[sm.cur_surface].max_aperture = 1.5
sm.gaps[sm.cur_surface].medium = med2   #   <-----   added
sm.ifcs[-1].max_aperture = 1.5
```

2. Use a single surface for the eye, with the image plane immersed in Medium Eye
```
#units in 'mm'
sm.gaps[0].thi=1e12#distant object
sm.add_surface([Rco, rp, n, 56., 1.5]) #This provides the surface representing the eye with a fake V-number
sm.gaps[sm.cur_surface].medium = med2 = ConstantIndex (n, 'Medium Eye') 
#---
sm.ifcs[-1].max_aperture = 1.5
```
ray-optics supports curved image surfaces so the second model is the most compact. Regardless of the approach adopted, the image gap medium must be the Medium Eye material, so that the image surface is considered immersed.

I think the spot diagram and aberration plots show that the image plane isn't at the circle of least confusion; it is at the paraxial focus. If you apply a defocus such that the real marginal ray is in focus (focus shift=-0.463653), then the spot diagram meets the criteria you calculated.
```
#Plot a Spot Diagram
#Fit the data to the same scale
spot_plt = plt.figure(FigureClass=SpotDiagramFigure, opt_model=opm, num_rays=64,
                      scale_type=Fit.All_Same, dpi=200, is_dark=isdark).plot()
```
![spotdgm](https://github.com/mjhoptics/ray-optics/assets/33294031/611d5aac-4374-48da-a725-27b599d13783)

Hope this helps.
Mike Hayford

**Comment by FERpo-23** [comment]
Hi, I really thank you your reply. It helped me a lot.
There are a couple of things that I do not catch up on completely: 
1.- The value for the "pupil" provided in <<osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=2.0*APER)>> is the diameter of the stop ? This is because I saw that you suggested a change in  <<sm.ifcs[sm.cur_surface].max_aperture = APER>> to half the value given in << PupilSpec >>
2.- Note that after making the change you suggested, the paraxial calculation for the image focal plane looks good, while the front focal plane is not properly determined (see the file enclosed).
3.- In addition, the lens layout looks pretty weird. Should I made any additional change ?

I will bite on the question of the aberrations later, when I fully understand the question in point 1.

Your work is highly appreciated. Regards.
[Reduced_eye_1S_V1a.tar.gz](https://github.com/mjhoptics/ray-optics/files/12342468/Reduced_eye_1S_V1a.tar.gz)

Fernando

**Reply by mjhoptics** [reply]
Hello Fernando,
Thank you for the response. Here are some additional details:
1) Optical design software traditionally has used the Entrance Pupil **Diameter** to specify the optical system aperture, while using the semi-diameter/radius for surface clear apertures. I blindly followed this convention to be consistent with CODE V, Zemax, and Oslo. I apologize for the confusion.
2) Thank you for noting this. I'll update a fix shortly. I've been researching the literature to try to make sure I'm using the correct sign convention, but there are minor differences between authors. I'll document what I use when I update the fix.
3) There are still a few issues with the lens layout; however, I have an alternative means of entering the data in terms of lens elements that renders correctly. Here's the setup:

```
#units in 'mm'

sm.gaps[0].thi=1e12 #distant object
med2 = ConstantIndex(n, 'Medium Eye')
eye_def = 1/Rco, 0., rp, med2, APER  # cv1, cv2, thi, medium, aperture
opm.add_lens(lens=eye_def)
#---
sm.gaps[-1].medium = med2   #   <-----   added
sm.ifcs[-1].max_aperture = APER
```

The layout then looks reasonable:
```
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
                        do_draw_rays=True, do_draw_ray_fans=True, do_paraxial_layout=False, is_dark=isdark).plot()
```
<img width="644" alt="Screenshot 2023-08-21 at 1 29 40 PM" src="https://github.com/mjhoptics/ray-optics/assets/33294031/6ea39c21-0bc3-48a0-8600-40072d611ed2">

Hope this helps,
Mike Hayford

**Reply by FERpo-23** [reply]
Hi Mike
I appreciate your answer.
You really helped me to understand the points I raised in my previous comments.
The question concerning aberrations is fully understandable.
Have a nice time!
Regards, Fernando
